### PR TITLE
MoveitCommander support for TrajectoryConstraints

### PR DESCRIPTION
**Pull request by bmagyar**
### Description

TLDR: 
* `trajectory_constraints` is a member of [`MotionPlanRequest`](http://docs.ros.org/kinetic/api/moveit_msgs/html/msg/MotionPlanRequest.html) yet the MoveGroupInterface doesn't allow for specifying it.
* `trajectory_constraints` can carry entire trajectories while `path_constraints` can't (in a clean way). 

----

The [`MotionPlanRequest`](http://docs.ros.org/kinetic/api/moveit_msgs/html/msg/MotionPlanRequest.html) message has a field `trajectory_constraints` but  `class MoveGroupInterface (C++)` doesn't allow for setting them. I've added support for this both to the C++ class and the Python wrapper. 
What's the motivation for this? Passing in _initial trajectories_, _initial guesses_ or _seed trajectories_ to optimization-based planners such as the case of [STOMP](https://github.com/ros-industrial/industrial_moveit/blob/kinetic-devel/stomp_moveit/src/stomp_planner.cpp#L505).

This was on #790 , now separated.

We should probably add some tests.

### Checklist
- [ ] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Optional: Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Optional: Decide if this should be cherry-picked to other current ROS branches (Indigo, Jade, Kinetic)

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by bmagyar** [comment]
Thanks for the excellent summary @v4hn ! Indeed, that is my interpretation of this field. Unfortunately even this doesn't really allow for a clean way of defining _multiple trajectories_. 

The definition of `moveit_msgs/Constraints` is as follows:
```
string name
moveit_msgs/JointConstraint[] joint_constraints
moveit_msgs/PositionConstraint[] position_constraints
moveit_msgs/OrientationConstraint[] orientation_constraints
moveit_msgs/VisibilityConstraint[] visibility_constraints
```

and while both `moveit_msgs/PositionConstraint` and `moveit_msgs/OrientationConstraint` allows to fully determine the full-DoF constraint of a given link, `moveit_msgs/JointConstraint` only addresses a single joint. This means that while a vector of Position and Orientaton constraints could determine a trajectory, a vector of PositionConstraints only determine a single joint-space waypoint.

Of course allowing for multiple Position and OrientationConstraints points beyond controlling a single point, which is a nice and general abstraction, we should keep it that way. 

I don't think this issue is critical enough to change message definitions but I'd like to mention that passing in trajectories could also be implemented by passing in a trajectory message but that implies adding a new field to `MotionPlanRequest`

File: moveit_commander/src/moveit_commander/move_group.py [hunk 1]
```diff
@@ -389,6 +390,27 @@ def clear_path_constraints(self):
...
 
+    def get_trajectory_constraints(self):
+        """ Get the actual trajectory constraints in form of a moveit_msgs.msgs.Constraints """
+        c = Constraints()
+        c_str = self._g.get_trajectory_constraints()
+        conversions.msg_from_string(c, c_str)
+        return c
+
+    def set_trajectory_constraints(self, value):
+        """ Specify the trajectory constraints to be used (as read from the database) """
```
Comment (rhaschke) [inline moveit_commander/src/moveit_commander/move_group.py]: Which database do you refer to?

File: moveit_commander/src/moveit_commander/move_group.py [hunk 2]
```diff
@@ -389,6 +390,27 @@ def clear_path_constraints(self):
...
+        if value == None:
+            self.clear_path_constraints()
```
Comment (rhaschke) [inline moveit_commander/src/moveit_commander/move_group.py]: Why this relays to path constraints, while the function is about trajectory constraints?

File: moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h [hunk 3]
```diff
@@ -906,6 +906,11 @@ class MoveGroupInterface
       This removes any path constraints set in previous calls to setPathConstraints(). */
   void clearPathConstraints();
 
+  moveit_msgs::TrajectoryConstraints getTrajectoryConstraints() const;
+  bool setTrajectoryConstraints(const std::string& constraints);
```
Comment (rhaschke) [inline moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h]: This variant doesn't seem to have an implementation. I think, it's superfluous.

File: moveit_commander/src/moveit_commander/move_group.py (hunk 1 above)
Comment (bmagyar) [reply moveit_commander/src/moveit_commander/move_group.py]: None actually. It's a ctrl+c&ctrl+v leftover

File: moveit_commander/src/moveit_commander/move_group.py (hunk 2 above)
Comment (bmagyar) [reply moveit_commander/src/moveit_commander/move_group.py]: It's a ctrl+c&ctrl+v leftover, fixed now

File: moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h (hunk 3 above)
Comment (bmagyar) [reply moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h]: Nice catch! Removed

**Comment by bmagyar** [comment]
> So far, the field is not used in any planner in MoveIt. 

It is. As mentioned above: [STOMP](https://github.com/ros-industrial/industrial_moveit/blob/kinetic-devel/stomp_moveit/src/stomp_planner.cpp#L505)

>  *  @bmagyar wants to use this field as a trajectory seed for planning, while the waypoints of the finally optimized trajectory might violate the "constraints".
> * @v4hn considered them as real constraints, i.e. the final solution should satisfy them with some waypoint in the given order.

Thank you @rhaschke for summing that up. 
We have precedent for both use-cases at the moment. While the bit of code from the [benchmark](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_ros/benchmarks/benchmarks/src/benchmark_execution.cpp#L450) that suggests `trajectory_constraints` are meant to be `waypoints` while [STOMP](https://github.com/ros-industrial/industrial_moveit/blob/kinetic-devel/stomp_moveit/src/stomp_planner.cpp#L505) uses it for `seed_trajectory`.

Short term: I think this should be merged into kinetic-devel. This is a fix for an interface that wasn't fully transparent with the C++ one before. My motivation for doing it is not relevant.

Long term: We discovered an ambiguity in the interface, we should aim to correct it. This deserves its own issue where we discuss the solution and then implement it. We should touch on this at the maintainers meeting!
My proposed solution:
 * deprecate `trajectory_constraints` from the message,
 * add `waypoints` or `waypoint_constraints`,
 * add `seed_trajectories`, `initial_trajectories`, `trajectory_guesses` -- we can agree on a name.

**Comment by v4hn** [comment]
I would agree that this can be merged without resolving the underlying problem.
We provided that interface before, it was just not wrapped.

Let's talk about the rest shortly next week, but feel free to open an issue already.

I'll leave it to @rhaschke to raise further discussion here or merge.
